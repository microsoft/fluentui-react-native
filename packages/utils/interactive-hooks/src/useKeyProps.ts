import * as React from 'react';
import { Platform } from 'react-native';

import { memoize } from '@fluentui-react-native/memo-cache';

import type { KeyCallback, KeyPressEvent, KeyPressProps } from './useKeyProps.types';

const shouldAllowShiftCtrlKeys = Platform.OS === ('win32' as any);

/**
 * Verifies if nativeEvent contains modifier key. The modifier keys that should
 * be taken into account differ based on platform
 * @param nativeEvent
 * @returns `true` if one or more of modifier keys are `true`
 */
const isModifierKey = (nativeEvent: any): boolean => {
  if (shouldAllowShiftCtrlKeys) {
    return nativeEvent && (nativeEvent.alt || nativeEvent.altKey || nativeEvent.meta || nativeEvent.metaKey);
  } else {
    return (
      nativeEvent &&
      (nativeEvent.alt ||
        nativeEvent.altKey ||
        nativeEvent.ctrl ||
        nativeEvent.ctrlKey ||
        nativeEvent.meta ||
        nativeEvent.metaKey ||
        nativeEvent.shift ||
        nativeEvent.shiftKey)
    );
  }
};

function keyPressCallback(userCallback?: KeyCallback, ...keys: string[]) {
  const onKeyEvent = (e: KeyPressEvent) => {
    if (userCallback !== undefined && !isModifierKey(e.nativeEvent) && (keys === undefined || keys.includes(e.nativeEvent.key))) {
      userCallback(e);
      e.stopPropagation();
    }
  };
  return onKeyEvent;
}

function getKeyUpPropsWorker(userCallback: KeyCallback, ...keys: string[]): KeyPressProps {
  const keyboardProps = Platform.select({
    ios: undefined,
    android: undefined,
    macos: {
      onKeyUp: keyPressCallback(userCallback, ...keys),
      validKeysUp: keys,
    },
    windows: {
      /**
       * https://github.com/microsoft/react-native-windows/issues/11049
       * Windows doesn't filter on `key` but on `code`, which is quite different ('A' vs 'KeyA' or 'GamepadA').
       * While this discrepancy is present, let's not specify `keyUpEvents`.
       */
      onKeyUp: keyPressCallback(userCallback, ...keys),
    },
    // win32
    default: {
      onKeyUp: keyPressCallback(userCallback, ...keys),
      keyUpEvents: keys.map((keyCode) => {
        return { key: keyCode };
      }),
    },
  });
  return keyboardProps;
}

function getKeyDownPropsWorker(userCallback: KeyCallback, ...keys: string[]): KeyPressProps {
  const keyboardProps = Platform.select({
    ios: undefined,
    android: undefined,
    macos: {
      onKeyDown: keyPressCallback(userCallback, ...keys),
      validKeysDown: keys,
    },
    windows: {
      /**
       * https://github.com/microsoft/react-native-windows/issues/11049
       * Windows doesn't filter on `key` but on `code`, which is quite different ('A' vs 'KeyA' or 'GamepadA').
       * While this discrepancy is present, let's not specify `keyDownEvents`.
       */
      onKeyDown: keyPressCallback(userCallback, ...keys),
    },
    // win32
    default: {
      onKeyDown: keyPressCallback(userCallback, ...keys),
      keyDownEvents: keys.map((keyCode) => {
        return { key: keyCode };
      }),
    },
  });
  return keyboardProps;
}

/**
 * Re-usable hook for an onKeyUp event.
 * @param userCallback The function you want to be called once the key has been activated on key up
 * @param keys A string of the key you want to perform some action on. If undefined, always invokes userCallback
 * @returns KeyPressProps: An object containing the correct platform specific props to handle key press
 */
export const useKeyUpProps = memoize(getKeyUpPropsWorker);

/**
 * Re-usable hook for an onKeyDown event.
 * @param userCallback The function you want to be called once the key has been activated on key down
 * @param keys A string of the key you want to perform some action on. If undefined, always invokes userCallback
 * @returns KeyPressProps: An object containing the correct platform specific props to  handle key press
 */
export const useKeyDownProps = memoize(getKeyDownPropsWorker);

/** Exposes the behavior of useKeyProps for the current platform as a boolean */
export const preferKeyDownForKeyEvents = Platform.select({
  macos: true,
  default: false,
});

/**
 * Re-usable hook for keyboard events. on macOS, this is onKeyDown, while on windows this is onKeyUp.
 * @param userCallback The function you want to be called once the key has been activated on key down
 * @param keys A string of the key you want to perform some action on. If undefined, always invokes userCallback
 * @returns KeyPressProps: An object containing the correct platform specific props to handle key press
 */
export const useKeyProps = preferKeyDownForKeyEvents ? useKeyDownProps : useKeyUpProps;

/**
 * Re-usable hook for an onKeyDown event.
 * @param userCallback The function you want to be called once the key has been activated on key up
 * @param keys A string of the key you want to perform some action on. If undefined, always invokes userCallback
 * @returns onKeyEvent() - Callback to determine if key was pressed, if so, call userCallback
 * @deprecated use the hook `useKeyProps` instead
 */
export function useKeyCallback(userCallback?: KeyCallback, ...keys: string[]) {
  const onKeyEvent = React.useCallback(
    (e: KeyPressEvent) => {
      if (userCallback !== undefined && (keys === undefined || keys.includes(e.nativeEvent.key))) {
        userCallback(e);
        e.stopPropagation();
      }
    },
    [keys, userCallback],
  );

  return onKeyEvent;
}
