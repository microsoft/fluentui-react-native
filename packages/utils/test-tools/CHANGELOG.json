{
  "name": "@fluentui-react-native/test-tools",
  "entries": [
    {
      "date": "Thu, 26 Aug 2021 04:48:40 GMT",
      "tag": "@fluentui-react-native/test-tools_v0.1.1",
      "version": "0.1.1",
      "comments": {
        "patch": [
          {
            "comment": "Bump @fluentui-react-native/theme-types to v0.11.3",
            "author": "67026167+chiuam@users.noreply.github.com",
            "commit": "b8c866644b2c5b64530990e59153424b0d6f5dcc",
            "package": "@fluentui-react-native/test-tools"
          }
        ]
      }
    },
    {
      "date": "Wed, 18 Aug 2021 23:00:57 GMT",
      "tag": "@fluentui-react-native/test-tools_v0.1.1",
      "version": "0.1.1",
      "comments": {
        "patch": [
          {
            "comment": "Bump @fluentui-react-native/theme-types to v0.11.2",
            "author": "ruaraki@microsoft.com",
            "commit": "345d1de930ff42359260da7f1e68a16059f4d763",
            "package": "@fluentui-react-native/test-tools"
          }
        ]
      }
    },
    {
      "date": "Fri, 13 Aug 2021 16:12:26 GMT",
      "tag": "@fluentui-react-native/test-tools_v0.1.1",
      "version": "0.1.1",
      "comments": {
        "patch": [
          {
            "comment": "Bump @fluentui-react-native/theme-types to v0.11.1",
            "author": "t-lindaweng@microsoft.com",
            "commit": "8946966df6ddc0ef1c0b43e576a4124c678d7c89",
            "package": "@fluentui-react-native/test-tools"
          }
        ]
      }
    },
    {
      "date": "Thu, 12 Aug 2021 17:23:44 GMT",
      "tag": "@fluentui-react-native/test-tools_v0.1.1",
      "version": "0.1.1",
      "comments": {
        "patch": [
          {
            "comment": "Bump @fluentui-react-native/theme-types to v0.11.0",
            "author": "ruaraki@microsoft.com",
            "commit": "e524662e3e40fe45a2dcc6e2acd3aca1197b5b5a",
            "package": "@fluentui-react-native/test-tools"
          }
        ]
      }
    },
    {
      "date": "Tue, 10 Aug 2021 21:12:30 GMT",
      "tag": "@fluentui-react-native/test-tools_v0.1.1",
      "version": "0.1.1",
      "comments": {
        "patch": [
          {
            "comment": "Bump @fluentui-react-native/theme-types to v0.10.0",
            "author": "ruaraki@microsoft.com",
            "commit": "ce49f33c08add7892191cf840d2fdcccfd0a16bb",
            "package": "@fluentui-react-native/test-tools"
          }
        ]
      }
    },
    {
      "date": "Sat, 07 Aug 2021 00:40:04 GMT",
      "tag": "@fluentui-react-native/test-tools_v0.1.1",
      "version": "0.1.1",
      "comments": {
        "none": [
          {
            "comment": "Bump @fluentui-react-native/theme-types to v0.9.3",
            "author": "jasonmo@microsoft.com",
            "commit": "e506337e109ef24e675c2ac794179db86fcb7e86",
            "package": "@fluentui-react-native/test-tools"
          }
        ]
      }
    },
    {
      "date": "Fri, 06 Aug 2021 00:36:02 GMT",
      "tag": "@fluentui-react-native/test-tools_v0.1.1",
      "version": "0.1.1",
      "comments": {
        "none": [
          {
            "comment": "Bump @fluentui-react-native/theme-types to v0.9.3",
            "author": "ruaraki@microsoft.com",
            "commit": "56ccce452d0dbbd5c24f53ebdd467f2d01dda694",
            "package": "@fluentui-react-native/test-tools"
          }
        ]
      }
    },
    {
      "date": "Wed, 04 Aug 2021 06:26:25 GMT",
      "tag": "@fluentui-react-native/test-tools_v0.1.1",
      "version": "0.1.1",
      "comments": {
        "patch": [
          {
            "comment": "Bump @fluentui-react-native/theme-types to v0.9.3",
            "author": "67026167+chiuam@users.noreply.github.com",
            "commit": "68dd61bd45ca8ffacc1580e468f3c01810e3093e",
            "package": "@fluentui-react-native/test-tools"
          }
        ]
      }
    },
    {
      "date": "Tue, 27 Jul 2021 22:17:20 GMT",
      "tag": "@fluentui-react-native/test-tools_v0.1.1",
      "version": "0.1.1",
      "comments": {
        "patch": [
          {
            "comment": "Bump @fluentui-react-native/theme-types to v0.9.2",
            "author": "jasonmo@microsoft.com",
            "commit": "1cd8611e1e88848ab66f7c0e9113c946f968260c",
            "package": "@fluentui-react-native/test-tools"
          }
        ]
      }
    },
    {
      "date": "Tue, 27 Jul 2021 18:55:58 GMT",
      "tag": "@fluentui-react-native/test-tools_v0.1.1",
      "version": "0.1.1",
      "comments": {
        "none": [
          {
            "comment": "Bump @fluentui-react-native/theme-types to v0.9.1",
            "author": "4123478+tido64@users.noreply.github.com",
            "commit": "07d04d166b13ff79e07a8c4cbc2e06afb1e4c3e6",
            "package": "@fluentui-react-native/test-tools"
          }
        ]
      }
    },
    {
      "date": "Mon, 26 Jul 2021 20:41:04 GMT",
      "tag": "@fluentui-react-native/test-tools_v0.1.1",
      "version": "0.1.1",
      "comments": {
        "patch": [
          {
            "comment": "Bump @fluentui-react-native/theme-types to v0.9.1",
            "author": "jasonmo@microsoft.com",
            "commit": "ad02fbbe9eaa5fbe6ec6126ccb719a1280f108a6",
            "package": "@fluentui-react-native/test-tools"
          }
        ]
      }
    }
  ]
}
