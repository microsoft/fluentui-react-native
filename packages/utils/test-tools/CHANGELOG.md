# Change Log - @fluentui-react-native/test-tools

This log was last generated on Thu, 26 Aug 2021 04:48:40 GMT and should not be manually modified.

<!-- Start content -->

## 0.1.1

Thu, 26 Aug 2021 04:48:40 GMT

### Patches

- Bump @fluentui-react-native/theme-types to v0.11.3 (67026167+chiuam@users.noreply.github.com)

## 0.1.1

Wed, 18 Aug 2021 23:00:57 GMT

### Patches

- Bump @fluentui-react-native/theme-types to v0.11.2 (ruaraki@microsoft.com)

## 0.1.1

Fri, 13 Aug 2021 16:12:26 GMT

### Patches

- Bump @fluentui-react-native/theme-types to v0.11.1 (t-lindaweng@microsoft.com)

## 0.1.1

Thu, 12 Aug 2021 17:23:44 GMT

### Patches

- Bump @fluentui-react-native/theme-types to v0.11.0 (ruaraki@microsoft.com)

## 0.1.1

Tue, 10 Aug 2021 21:12:30 GMT

### Patches

- Bump @fluentui-react-native/theme-types to v0.10.0 (ruaraki@microsoft.com)

## 0.1.1

Wed, 04 Aug 2021 06:26:25 GMT

### Patches

- Bump @fluentui-react-native/theme-types to v0.9.3 (67026167+chiuam@users.noreply.github.com)

## 0.1.1

Tue, 27 Jul 2021 22:17:20 GMT

### Patches

- Bump @fluentui-react-native/theme-types to v0.9.2 (jasonmo@microsoft.com)

## 0.1.1

Mon, 26 Jul 2021 20:41:04 GMT

### Patches

- Bump @fluentui-react-native/theme-types to v0.9.1 (jasonmo@microsoft.com)
