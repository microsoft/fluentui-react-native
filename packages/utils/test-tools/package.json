{
  "name": "@fluentui-react-native/test-tools",
  "version": "0.1.1",
  "description": "Tools and mocks for testing components using jest and enzyme",
  "main": "lib-commonjs/index.js",
  "module": "lib/index.js",
  "typings": "lib/index.d.ts",
  "private": true,
  "scripts": {
    "build": "fluentui-scripts build",
    "just": "fluentui-scripts",
    "clean": "fluentui-scripts clean",
    "depcheck": "fluentui-scripts depcheck",
    "lint": "fluentui-scripts eslint",
    "test": "fluentui-scripts jest",
    "update-snapshots": "fluentui-scripts jest -u",
    "verify-api": "fluentui-scripts verify-api-extractor",
    "update-api": "fluentui-scripts update-api-extractor"
  },
  "dependencies": {
    "@fluentui-react-native/theme-types": ">=0.3.0 <1.0.0"
  },
  "devDependencies": {
    "@types/react": "^16.9.34",
    "@types/react-native": "^0.62.0",
    "@uifabricshared/build-native": "^0.1.1",
    "@uifabricshared/eslint-config-rules": "^0.1.1",
    "react": "16.11.0",
    "react-native": "0.62.2"
  },
  "peerDependencies": {
    "react": "^16.8.0",
    "react-native": ">=0.60.0"
  },
  "author": "",
  "license": "MIT"
}
