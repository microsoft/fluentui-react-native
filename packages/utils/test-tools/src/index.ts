export * from './enzymeTests';
export * from './mockTheme';
