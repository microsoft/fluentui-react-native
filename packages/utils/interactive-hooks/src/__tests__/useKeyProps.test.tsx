import * as React from 'react';
import { Pressable } from 'react-native';

import { checkRenderConsistency, checkReRender } from '@fluentui-react-native/test-tools';
import * as renderer from 'react-test-renderer';

import { useKeyProps } from '../useKeyProps';
import type { PressablePropsExtended } from '../usePressableState.types';

const dummyFunction = () => {
  console.log('dummy');
};

// Simple wrapper function to let us use `PressablePropsExtended` to fix type errors
const PressableWithDesktopProps = (props: PressablePropsExtended) => {
  return <Pressable {...props} />;
};

it('Pressable with useKeyProps', () => {
  const TestComponent = () => {
    const keyboardProps = useKeyProps(dummyFunction, ' ', 'Enter');
    return <PressableWithDesktopProps {...keyboardProps} />;
  };

  const tree = renderer.create(<TestComponent />).toJSON();
  expect(tree).toMatchSnapshot();
});

it('useKeyProps called twice', () => {
  const TestComponent = () => {
    const keyboardProps1 = useKeyProps(dummyFunction, ' ', 'Enter');
    const keyboardProps2 = useKeyProps(dummyFunction, ' ', 'Enter');
    expect(keyboardProps1).toBe(keyboardProps2);
    return <PressableWithDesktopProps {...keyboardProps2} />;
  };

  const tree = renderer.create(<TestComponent />).toJSON();
  expect(tree).toMatchSnapshot();
});

it('Pressable with useKeyProps simple rendering does not invalidate styling', () => {
  const TestComponent = () => {
    const keyboardProps = useKeyProps(dummyFunction, ' ', 'Enter');
    return <PressableWithDesktopProps {...keyboardProps} />;
  };

  checkRenderConsistency(() => <TestComponent />, 2);
});

it('Pressable with useKeyProps re-renders correctly', () => {
  const TestComponent = () => {
    const keyboardProps = useKeyProps(dummyFunction, ' ', 'Enter');
    return <PressableWithDesktopProps {...keyboardProps} />;
  };
  checkReRender(() => <TestComponent />, 2);
});
