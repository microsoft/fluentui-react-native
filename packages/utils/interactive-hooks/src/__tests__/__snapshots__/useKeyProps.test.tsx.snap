// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pressable with useKeyProps 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "multiselectable": undefined,
      "required": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  focusable={true}
  keyUpEvents={
    [
      {
        "key": " ",
      },
      {
        "key": "Enter",
      },
    ]
  }
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onKeyDown={[Function]}
  onKeyUp={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
/>
`;

exports[`useKeyProps called twice 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "multiselectable": undefined,
      "required": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  focusable={true}
  keyUpEvents={
    [
      {
        "key": " ",
      },
      {
        "key": "Enter",
      },
    ]
  }
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onKeyDown={[Function]}
  onKeyUp={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
/>
`;
